README.MD
![Alt text](https://raw.githubusercontent.com/JoemarDev/School-System-API/main/logo.png)
# Schooolify API
Schooolify api provides you a ready made API System for mostly used in school management system.

* API BETA Release(v1.0.1)
* You can check the (Prototype) API Documention in this link [Schooolify API](https://documenter.getpostman.com/view/16604401/2s7Ymz8Ma7).

## Change Log
[Change log records](https://github.com/JoemarDev/Schooolyfi-API/blob/main/CHANGELOG.MD)

## Installation
Install Packages
```bash
npm install
```
And Run
```bash
npm run dev
```

For Creating a dummy data run
```bash
node PopulateData.js
```

For Resetting the database run
```bash
node ResetDatabase.js
```

## Features
1. Authentication by jsonwebtoken and implemented via cookies. (3 types of role)
    - Admin Account
    - Student Account
    - Teacher Account
2. Student Mangement
3. Teacher Management
4. Lesson Plan Management
5. Subjects Management
6. Course Management
7. Schedule Management
8. Grading Calculation

## Future Plan
1. Add Parent Portal
2. Add Online Classroom
    - Classroom Chat
    - Video class room.


## TECH
* [NODE JS] - ES6
* [MONGODB] - NOSQL


## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

## License
[GNU General Public License v3.0](https://github.com/JoemarDev/Schooolyfi-API/blob/main/LICENSE)

CHANGELOG.MD
![Alt text](https://raw.githubusercontent.com/JoemarDev/School-System-API/main/logo.png)
# Schooolify API

* This API is still on progress
* You can check the (Prototype) API Documention in this link [Schooolify API](https://documenter.getpostman.com/view/16604401/2s7Ymz8Ma7).


## Change Logs

### v0.0.1 (09/14/22)
1. Create Schema Model for Student , Teacher , Subject 
2. Create Controller for the models
3. Add CRUD Functionality in controllers.


### v0.0.2 (09/15/22)
1. Add Authentication for teacher and students (Cookies , JWT , Middleware)
2. Teacher routes is not allowed for students cookies
3. Add Profile Routes for Teacher and Student 
    - Able to see User Information
4. Able to assign a subject to teacher
    - Check if teacher is applicable to specific schedule
    - Compute the Class Session duration
5. Student Is able to pick a subject schedule
    - Student can look up to the subject list with the teacher who is teaching it.
    - Time compatible able to anylyze if the student is fit on that schedule.
6. Student is able to see his schedule 
    - Subject Info
    - Teacher Info


### v0.0.3 (09/16/22)
1. Refractor Student Schedule and Subject Schedule Controller and convert it to One Controller as ScheduleController
2. Able to remove the subject schedule from student.
3. Able to change the subject schedule 
    - Checks if it will have conflict to other subject
    - Pick from the subject schedule that already assign from the teacher.
4. Clean then authentication middlewares in route files. 
    - Add All Persmision to middleware to temporary access all functionality whoever is login
5. Add Routes to get student schedule information
6. New Models 
    - Quiz
    - Exam,

### v0.0.4 (09/17/22)
1. Get teacher Schedule
2. Get all the student list from the subject that is scheduled to teacher.
3. Update the Create Student Subject Schedule 
    - from receving the student from the cookie
    - to manually add it from body.
4. Able to remove the subject to teacher
    - Remove Also From The Student Schedule
5. Able to update the teacher from shedule


### v0.0.5 (09/18/22)
1. Add .env_example
2. Course
    - Create a model for COURSE
    - CRUD IMPLEMENTATION
    - Course must have a subject list from (SUBJECT MODEL)
    - Able to get all the student have this course
3. Subject Model 
    - Must have a lesson plan (Lesson Plan Model)
    - Able to show all the lesson plan for the subjects.
4. Lesson Plan
    - Create a model for LessonPlan
    - CRUD IMPLEMENTATION
5. Able to check all the students Subjects


### v0.0.6 (09/19/22)
1. Remove Quiz model 
    - Rerfractor , Join it in Exam Model.
2. Implement Exam CRUD
    - Able to automatically compute the average of the exams
    - exam can only be add based on the subject and the course of students as well quiz.
    - add enum attributes (exam, quiz)
3. Combine Exam and Quiz Controller 
    - Add Middleware in the routes to identify the type of exam


### v0.0.7 (09/20/22)
1. Implement Attendace 
    - CRUD Features
    - Student must have only 1 attendace perday
    - attendance must be under in the subject
    - check if the student is own the subject
2. Create a route where it can calculate the average attendance of the student
3. Implement Project & Activity
    - CRUD Implementation
    - Subject Student Course Validator
4. Check student Subject Average


### v0.0.8 (09/21/22)
1. Implement Grading Formula
    - CRUD Functionality
    - Fixed Amount of Grading Fomula 
        Allowed :
        - quiz
        - attendace 
        - exam
        - project
        - activity
    - Automaticlly update the existing one , if create with the same type of fomula
2. Getting the average of the student subject is based on Grading Formula
3. Check Student Total Average


### v0.0.9 (09/22/22)

1. Work on progress populate script
2. Re-work the codes  of checking if the subject have the course
3. Change course data schema type of subject.
4. Update the ResetDatabase Script
5. Update the pre methods in Subject Model
    - Delete all related model when the subject is deleted.

### v1.0.0 (09/24/22)
1. Update The Controller of creating subject , make sure that it validate the course that will be passed
    - If the course is passed as a string convert it to array to properly handle in the mongo schema


### v1.0.1(09/25/22)
1. Create Admin Account
2. Create Route For Creating a admin account
    - Check if have a existing admin account
    - Create Route for creating admin account
    - Can update admin password
3. Create a authentication route for admin
4. Create a populate script
5. Finish Reset Database Script


### TODO for v1.0.2
1. Create a test script
2. Create a swagger UI documentation
3. Create a README Installation
    - Provide the features of the system
    - Provide the future feature to be realease.
4. Make the first realease.   

PopulateData.js
const StudentData = require('./Mock-Data/Student.json');
const TeacherData = require('./Mock-Data/Teacher.json');
const CourseData = require('./Mock-Data/Courses.json');
const SubjectData = require('./Mock-Data/Subject.json');
const Student = require('./src/model/Student');
const Teacher = require('./src/model/Teacher');
const Course = require('./src/model/Course');
const Subject = require('./src/model/Subject');
const SubjecScheduleSchema = require('./src/model/Subject-Schedule');
const StudentSchedule = require('./src/model/Student-Schedule');
const SubjectSchedule = require('./src/model/Subject-Schedule');
require('dotenv').config();

// Database
const connectDB = require('./src/database/connection');

// Populate Student
const CreateStudent = async() => {
    for(const student of  StudentData) {

        student.password = '1234';
        
        const CourseCount = await Course.countDocuments();
        
        let randomCourse = Math.floor(Math.random() * CourseCount);

        const course = await Course.findOne().skip(randomCourse);

        student.course = course;

        await Student.create(student);
    }
    console.log(`Add ${StudentData.length} Students.`);
};

// Populate Teacher
const CreateTeacher = async() => {
    for(const teacher of  TeacherData) {
        teacher.password = '1234';
        await Teacher.create(teacher);
    }
    console.log(`Add ${StudentData.length} Teacher.`);
};


// Populate Course
const CreateCourse = async() => {
    for(const course of CourseData) {
        await Course.create(course);
    }
    console.log(`Add ${CourseData.length} Course.`);
}


const CreateSubject = async() => {

    const courses = await Course.find({}).select('_id');

    for(const subject of SubjectData) {
        subject.courses = courses;
        await Subject.create(subject);
    }

    console.log(`Add ${CourseData.length} Subject.`);
}


const AssignScheduleToTeacher = async() => {
    
    const teachers = await Teacher.find({});
    const subjects = await Subject.find({});
    
    const ScheduleLists = [
        {
            classTimeStart : "13:00",
            classTimeDismiss : "14:30",
            classTotalConsumeTime : 30,
        },
        {
            classTimeStart : "14:20",
            classTimeDismiss : "15:50",
            classTotalConsumeTime : 70,
        },
        {
            classTimeStart : "16:10",
            classTimeDismiss : "17:50",
            classTotalConsumeTime : 60,
        },
        {
            classTimeStart : "18:10",
            classTimeDismiss : "19:50",
            classTotalConsumeTime : 60,
        }
    ];

    let scheduleList  = [];


    for(const teacher of teachers) {
        for(const subject of subjects) {

            let randomSched  = ScheduleLists[Math.floor(Math.random() * ScheduleLists.length)]
            let data = {
                    "teacher" : teacher._id,
                    "subject" : subject._id,
                    "classTimeStart" : randomSched['classTimeStart'],
                    "classTimeDismiss" : randomSched['classTimeDismiss'],
                    "classTotalConsumeTime" : randomSched['classTotalConsumeTime']
                }

            scheduleList = [...scheduleList , data];
        }
    }

    await SubjecScheduleSchema.insertMany(scheduleList);

    console.log(`Inserted ${scheduleList.length} Teacher Schedule`);
}

const CreateStudentSchedule = async() => {
    const students = await Student.find({});

    let scheduleList = [];
    for(student of students) {
        const subjects = await Subject.find({course : student.course});
        for(const subject of subjects) {
            const availableTeacher = await SubjectSchedule.find({subject : subject._id});
            const RandomTeacher = Math.floor(Math.random() * availableTeacher.length);
            let data = {
                "subjectSchedule": availableTeacher[RandomTeacher]._id,
                "student" : student._id
            }
            scheduleList = [...scheduleList , data]
        }
    }

    await StudentSchedule.insertMany(scheduleList);

    console.log(`Inserted ${scheduleList.length} Student Schedule`);

};

const Start = async() => {
    await connectDB(process.env.MONGO_URI);    
    await CreateCourse();
    await CreateTeacher();
    await CreateStudent();
    await CreateSubject();
    await AssignScheduleToTeacher();
    await CreateStudentSchedule();
    process.exit(0);
};

Start();